enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/EagerRegisterPipe.scala

```diff
@@ -98,12 +101,42 @@ case class EagerRegisterPipe(source: Pipe, pipelineInformation: PipelineInformat
           private def fail(): Nothing =
             throw new InternalException(s"Not supported in anonymous ${classOf[EagerRegisterPipe]} execution context")
         }
-        primitiveIndex += primitiveColumnCount
+        longIndex += longColumnCount
         refIndex += refColumnCount
 
         row
       }
     }
   }
 
+  private class ResizableLongArray(initialSize: Int, growth: Int) extends ResizableArray[Long](initialSize, growth) {
+    override protected def newArray(size: Int): Array[Long] = new Array[Long](size)
+  }
+
+  private class ResizableRefArray(initialSize: Int, growth: Int) extends ResizableArray[Any](initialSize, growth) {
+    override protected def newArray(size: Int): Array[Any] = new Array[Any](size)
+  }
+
+  private abstract class ResizableArray[T](initialSize: Int, growth: Int) {
+    var array: Array[T] = newArray(initialSize)
+
+    var length: Int = 0
+
+    def ++=(elements: Array[T]): Unit = {
+      while (length + elements.length > array.length)
+        grow()
+      System.arraycopy(elements, 0, array, length, elements.length)
+      length = length + elements.length
+    }
+
+    protected def newArray(size: Int): Array[T]
+
+    // NOTE: current impl doesn't reuse space of old array. Instead, it creates bigger array & copies old elements to it
```

Comment:
@systay @eebus if/when reviewing PR please pay special attention to `ResizableArray`.
It may turn out you want to revert this commit, to go back to using `ArrayBuffer` (FYI, all tests still pass if this commit is reverted).
The benefit of aggregating into `Array` is we get to `arraycopy` in many places, which (presumably) becomes more beneficial with more columns.
Downside of current impl is (as in comment) every time we resize array we *copy* elements to new array. It's done with `arraycopy` so hopefully low cost, but it *does* mean that we waste space (and some time). 
In worst case this impl consumes 3x more space than necessary -- if we reused old array and "appended arrays" (as in your PR, Andres) then worst case would be 2x (when we are one element past resize border).